Update of the International Geomagnetic Reference Field from Version 12 (default) to Version 13. Updated coefficients
in the g1 parameter cofrm subroutine (provided in Alken 2021 paper), and updated write dates.
Intended for use up to 2025, but warnings about using a future geomagnetic field are written to the output file from post-2023.

Author: Matthew Brown, Space Environment & Radio Engineering Research Group (SERENE), University of Birmingham.

diff -Nru tiegcm2.0.orig/src/apex.F90 tiegcm2.0/src/apex.F90
--- tiegcm2.0.orig/src/apex.F90	2022-10-25 17:47:01.436862000 +0100
+++ tiegcm2.0/src/apex.F90	2022-10-25 17:50:52.053751000 +0100
@@ -1864,19 +1864,19 @@
   integer,parameter :: n1=120, n2=195, isv=0
   integer,parameter :: &
     ncn1=19, & ! number of coefficients dimensioned n1
-    ncn2=6     ! number of coefficients dimensioned n2: increase with each IGRF update
+    ncn2=7     ! number of coefficients dimensioned n2: increase with each IGRF update
   integer,parameter :: ngh = n1*ncn1 + n2*ncn2 + 1 ! not sure why the extra +1
   real,save :: g1(n1,ncn1), g2(n2,ncn2), gh(ngh)
   real,parameter :: alt = 0.
 
-  if (date < 1900. .or. date > 2025.) then
-    write(6,"('>>> cofrm: date=',f8.2,' Date must be >= 1900 and <= 2025')") date
+  if (date < 1900. .or. date > 2030.) then
+    write(6,"('>>> cofrm: date=',f8.2,' Date must be >= 1900 and <= 2030')") date
     stop 'cofrm'
   endif
-  if (date > 2020.) then
+  if (date > 2025.) then
     write(6,"('>>> WARNING cofrm:')")
     write(6,"(/,'   This version of IGRF is intended for use up to ')")
-    write(6,"('     2020. Values for ',f9.3,' will be computed but')") date
+    write(6,"('     2025. Values for ',f9.3,' will be computed but')") date
     write(6,"('     may be of reduced accuracy.',/)")
   endif
 
@@ -2325,49 +2325,79 @@
     	 -0.53,   -0.26,   -0.79/)			  	
   
    g2(:,5) = (/                                               & ! 2015
-    -29442.0,-1501.0, 4797.1,-2445.1, 3012.9,-2845.6, 1676.7, &  
-      -641.9, 1350.7,-2352.3, -115.3, 1225.6,  244.9,  582.0, &   
-      -538.4,  907.6,  813.7,  283.3,  120.4, -188.7, -334.9, &   
-       180.9,	70.4, -329.5, -232.6,  360.1,	47.3,  192.4, &   
-       197.0, -140.9, -119.3, -157.5,	16.0,	 4.1,  100.2, &   
-    	70.0,	67.7,  -20.8,	72.7,	33.2, -129.9,	58.9, &   
-       -28.9,  -66.7,	13.2,	 7.3,  -70.9,	62.6,	81.6, &   
-       -76.1,  -54.1,	-6.8,  -19.5,	51.8,	 5.7,	15.0, &   
-    	24.4,	 9.4,	 3.4,	-2.8,  -27.4,	 6.8,	-2.2, &   
-    	24.2,	 8.8,	10.1,  -16.9,  -18.3,	-3.2,	13.3, &   
-       -20.6,  -14.6,	13.4,	16.2,	11.7,	 5.7,  -15.9, &   
-    	-9.1,	-2.0,	 2.1,	 5.4,	 8.8,  -21.6,	 3.1, &   
-    	10.8,	-3.3,	11.8,	 0.7,	-6.8,  -13.3,	-6.9, &   
-    	-0.1,	 7.8,	 8.7,	 1.0,	-9.1,	-4.0,  -10.5, &   
-    	 8.4,	-1.9,	-6.3,	 3.2,	 0.1,	-0.4,	 0.5, &   
-    	 4.6,	-0.5,	 4.4,	 1.8,	-7.9,	-0.7,	-0.6, &   
-    	 2.1,	-4.2,	 2.4,	-2.8,	-1.8,	-1.2,	-3.6, &   
-    	-8.7,	 3.1,	-1.5,	-0.1,	-2.3,	 2.0,	 2.0, &   
-    	-0.7,	-0.8,	-1.1,	 0.6,	 0.8,	-0.7,	-0.2, &   
-    	 0.2,	-2.2,	 1.7,	-1.4,	-0.2,	-2.5,	 0.4, &   
-    	-2.0,	 3.5,	-2.4,	-1.9,	-0.2,	-1.1,	 0.4, &   
-    	 0.4,	 1.2,	 1.9,	-0.8,	-2.2,	 0.9,	 0.3, &   
-    	 0.1,	 0.7,	 0.5,	-0.1,	-0.3,	 0.3,	-0.4, &   
-    	 0.2,	 0.2,	-0.9,	-0.9,	-0.1,	 0.0,	 0.7, &   
-    	 0.0,	-0.9,	-0.9,	 0.4,	 0.4,	 0.5,	 1.6, &   
-    	-0.5,	-0.5,	 1.0,	-1.2,	-0.2,	-0.1,	 0.8, &   
-    	 0.4,	-0.1,	-0.1,	 0.3,	 0.4,	 0.1,	 0.5, &   
-    	 0.5,	-0.3,	-0.4,	-0.4,	-0.3,	-0.8/)	         
+    -29441.5,-1501.8, 4796.0,-2445.9, 3012.2,-2845.4, 1676.4, &  
+      -642.2, 1350.3,-2352.3, -115.3, 1225.9,  245.0,  581.7, &   
+      -538.7,  907.4,  813.7,  283.5,  120.5, -188.4, -334.9, &   
+       181.0,	70.4, -329.2, -232.9,  360.1,	47.0,  192.4, &   
+       197.0, -140.9, -119.1, -157.4,	16.0,	 4.3,  100.1, &   
+    	70.0,	67.6,  -20.6,	72.8,	33.3, -129.9,	58.7, &   
+       -28.9,  -66.6,	13.1,	 7.4,  -70.9,	62.4,	81.3, &   
+       -76.0,  -54.3,	-6.8,  -19.5,	51.8,	 5.6,	15.1, &   
+    	24.5,	 9.3,	 3.3,	-2.9,  -27.5,	 6.6,	-2.3, &   
+    	24.0,	 8.9,	10.0,  -16.8,  -18.3,	-3.2,	13.2, &   
+       -20.6,  -14.6,	13.3,	16.2,	11.8,	 5.7,  -16.0, &   
+    	-9.1,	-2.0,	 2.3,	 5.3,	 8.8,  -21.8,	 3.0, &   
+    	10.8,	-3.2,	11.7,	 0.7,	-6.7,  -13.2,	-6.9, &   
+    	-0.1,	 7.8,	 8.7,	 1.0,	-9.1,	-3.9,  -10.5, &   
+    	 8.4,	-2.0,	-6.3,	 3.3,	 0.2,	-0.4,	 0.6, &   
+    	 4.6,	-0.6,	 4.4,	 1.7,	-7.9,	-0.7,	-0.6, &   
+    	 2.1,	-4.2,	 2.3,	-2.9,	-1.8,	-1.1,	-3.6, &   
+    	-8.7,	 3.0,	-1.4,	-0.0,	-2.3,	 2.1,	 2.1, &   
+    	-0.6,	-0.8,	-1.1,	 0.6,	 0.8,	-0.7,	-0.2, &   
+    	 0.1,	-2.1,	 1.7,	-1.4,	-0.2,	-2.6,	 0.4, &   
+    	-2.0,	 3.5,	-2.3,	-2.1,	-0.2,	-1.1,	 0.5, &   
+    	 0.4,	 1.2,	 1.8,	-0.9,	-2.2,	 0.9,	 0.3, &   
+    	 0.1,	 0.7,	 0.5,	-0.1,	-0.4,	 0.3,	-0.4, &   
+    	 0.2,	 0.2,	-0.9,	-0.9,	-0.2,	 0.0,	 0.7, &   
+    	 0.0,	-0.9,	-0.9,	 0.4,	 0.5,	 0.6,	 1.6, &   
+    	-0.4,	-0.5,	 1.0,	-1.2,	-0.2,	-0.1,	 0.8, &   
+    	 0.4,	-0.1,	 0.0,	 0.4,	 0.5,	 0.1,	 0.5, &   
+    	 0.5,	-0.3,	-0.4,	-0.4,	-0.4,	-0.7/)	         
+  
+   g2(:,6) = (/                                               & ! 2020
+    -29404.8,-1450.9, 4652.5,-2499.6, 2982.0,-2991.6, 1677.0, &  
+      -734.6, 1363.2,-2381.2,  -82.1, 1236.2,  241.9,  525.7, &   
+      -543.4,  903.0,  809.5,  281.9,   86.3, -158.4, -309.4, &   
+       199.7,	48.0, -349.7, -234.3,  363.2,	47.7,  187.8, &   
+       208.3, -140.7, -121.2, -151.2,	32.3,	13.5,   98.9, &   
+    	66.0,	65.5,  -19.1,	72.9,	25.1, -121.5,	52.8, &   
+       -36.2,  -64.5,	13.5,	 8.9,  -64.7,	68.1,	80.6, &   
+       -76.7,  -51.5,	-8.2,  -16.9,	56.5,	 2.2,	15.8, &   
+    	23.5,	 6.4,	-2.2,	-7.2,  -27.2,	 9.8,	-1.8, &   
+    	23.7,	 9.7,	 8.4,  -17.6,  -15.3,	-0.5,	12.8, &   
+       -21.1,  -11.7,	15.3,	14.9,	13.7,	 3.6,  -16.5, &   
+    	-6.9,	-0.3,	 2.8,	 5.0,	 8.4,  -23.4,	 2.9, &   
+    	11.0,	-1.5,	 9.8,	-1.1,	-5.1,  -13.2,	-6.3, &   
+    	 1.1,	 7.8,	 8.8,	 0.4,	-9.3,	-1.4,  -11.9, &   
+    	 9.6,	-1.9,	-6.2,	 3.4,	-0.1,	-0.2,	 1.7, &   
+    	 3.6,	-0.9,	 4.8,	 0.7,	-8.6,	-0.9,	-0.1, &   
+    	 1.9,	-4.3,	 1.4,	-3.4,	-2.4,	-0.1,	-3.8, &   
+    	-8.8,	 3.0,	-1.4,	-0.0,	-2.5,	 2.5,	 2.3, &   
+    	-0.6,	-0.9,	-0.4,	 0.3,	 0.6,	-0.7,	-0.2, &   
+    	-0.1,	-1.7,	 1.4,	-1.6,	-0.6,	-3.0,	 0.2, &   
+    	-2.0,	 3.1,	-2.6,	-2.0,	-0.1,	-1.2,	 0.5, &   
+    	 0.5,	 1.3,	 1.4,	-1.2,	-1.8,	 0.7,	 0.1, &   
+    	 0.3,	 0.8,	 0.5,	-0.2,	-0.3,	 0.6,	-0.5, &   
+    	 0.2,	 0.1,	-0.9,	-1.1,	 0.0,	-0.3,	 0.5, &   
+    	 0.1,	-0.9,	-0.9,	 0.5,	 0.6,	 0.7,	 1.4, &   
+    	-0.3,	-0.4,	 0.8,	-1.3,	 0.0,	-0.1,	 0.8, &   
+    	 0.3,	 0.0,	-0.1,	 0.4,	 0.5,	 0.1,	 0.5, &   
+    	 0.5,	-0.4,	-0.5,	-0.4,	-0.4,	-0.6/)	         
 !
-    g2(1:80,6) = (/                                           & ! 2017
-       10.3,   18.1,  -26.6,   -8.7,   -3.3,  -27.4,    2.1,  &  
-      -14.1,	3.4,   -5.5,	8.2,   -0.7,   -0.4,  -10.1,  &   
-     	1.8,   -0.7,	0.2,   -1.3,   -9.1,	5.3,	4.1,  &   
-     	2.9,   -4.3,   -5.2,   -0.2,	0.5,	0.6,   -1.3,  &   
-     	1.7,   -0.1,   -1.2,	1.4,	3.4,	3.9,	0.0,  &   
-       -0.3,   -0.1,	0.0,   -0.7,   -2.1,	2.1,   -0.7,  &   
-       -1.2,	0.2,	0.3,	0.9,	1.6,	1.0,	0.3,  &   
-       -0.2,	0.8,   -0.5,	0.4,	1.3,   -0.2,	0.1,  &   
-       -0.3,   -0.6,   -0.6,   -0.8,	0.1,	0.2,   -0.2,  &   
-     	0.2,	0.0,   -0.3,   -0.6,	0.3,	0.5,	0.1,  &   
-       -0.2,	0.5,	0.4,   -0.2,	0.1,   -0.3,   -0.4,  &   
-     	0.3,	0.3,	0.0/)
-    g2(81:n2,6) = 0.0
+    g2(1:80,7) = (/                                           & ! 2022
+        5.7,    7.4,  -25.9,  -11.0,   -7.0,  -30.2,   -2.1,  &  
+      -22.4,	2.2,   -5.9,	6.0,    3.1,   -1.1,  -12.0,  &   
+     	0.5,   -1.2,   -1.6,   -0.1,   -5.9,	6.5,	5.2,  &   
+     	3.6,   -5.1,   -5.0,   -0.3,	0.5,	0.0,   -0.6,  &   
+     	2.5,    0.2,   -0.6,	1.3,	3.0,	0.9,	0.3,  &   
+       -0.5,   -0.3,	0.0,    0.4,   -1.6,	1.3,   -1.3,  &   
+       -1.4,	0.8,	0.0,	0.0,	0.9,	1.0,   -0.1,  &   
+       -0.2,	0.6,    0.0,	0.6,	0.7,   -0.8,	0.1,  &   
+       -0.2,   -0.5,   -1.1,   -0.8,	0.1,	0.8,    0.3,  &   
+     	0.0,	0.1,   -0.2,   -0.1,	0.6,	0.4,   -0.2,  &   
+       -0.1,	0.5,	0.4,   -0.3,	0.3,   -0.4,   -0.1,  &   
+     	0.5,	0.4,	0.0/)
+    g2(81:n2,7) = 0.0
 !
 ! Set gh from g1,g2:
 !
@@ -2383,7 +2413,7 @@
   enddo
   gh(ngh) = 0. ! not sure why gh is dimensioned with the extra element, so set it to 0. 
   
-  if (date < 2015.) then
+  if (date < 2020.) then
     t   = 0.2*(date - 1900.)
     ll  = t
     one = ll
@@ -2405,8 +2435,8 @@
       tc = -0.2
       t = 0.2
     endif
-  else ! date >= 2015
-    t     = date - 2015.0
+  else ! date >= 2020
+    t     = date - 2020.0
     tc    = 1.0
     if (isv.eq.1) then
       t = 1.0
@@ -2416,7 +2446,7 @@
     nmx   = 13
     nc    = nmx*(nmx+2)
     kmx   = (nmx+1)*(nmx+2)/2
-  endif ! date < 2015
+  endif ! date < 2020
   r = alt
   l = 1
   m = 1
